### ISSUE TITLE
Easy way to run only certain test suites / test cases with `raco test`

### ISSUE THREAD

**Issue by sschwarzer** [racket/racket#4107]
**Is your feature request related to a problem? Please describe.**
Occasionally, for example when changing internal APIs, I run into the situation that a good part of my existing test cases temporarily break because they don't test the new API yet. In such a case, it would be helpful to run only a part of the tests to avoid being overwhelmed by a lot of failing tests and to focus first on the most "basic" tests affected by the API change.

**Describe the solution you'd like**
I usually use `raco test` and `rackunit` for my tests. It would be great if there was a way to select only certain tests (test suites / test cases) from the command line. Since it's already possible to assign descriptions to test suites and test cases, it would be natural if the command line API took the description strings or regex patterns for them. Also, there could be `raco test` flags to _include_ or _exclude_ tests with the given descriptions.

**Describe alternatives you've considered**
In a case as described above, I work through the files one by one, starting with the most immediately affected module. Inside the module, I comment out the tests I currently don't want to run with `#| ... |#`.

**Do you want to contribute to this feature**
I'd like to, but at the moment I'm also working on other Racket stuff, so I don't know when I'll have the time. I'm not offended if someone else wants to work on this feature, but in that case the person should please add a note to this ticket. :-)

**Additional context**
- I'm not sure yet how practical the test selection based on the test suite/case descriptions is going to be.

- I brought up the issue on the Racket Slack and @samth suggested this implementation approach:

  *[...] I think a good approach would be (a) having raco test take an argument that sets some parameter that (b) rackunit or other testing frameworks could read and decide to run/not run some tests.*

**Comment by mfelleisen** [comment]
I no longer use test suites but collect tests in submodules. That way I can run `raco test -s SM` for just the set of tests I want.  I don't consider this a work-around; I think rackunit (and perhaps unit testing in general) is too complicated as-is and that submodules solve a number of the problems. (This may sound strange coming from someone who always looks for linguistic solutions for a tool-chain solution but I have also come to see some `raco` as part of my language. YMMD)

**Comment by sschwarzer** [comment]
@mfelleisen Thanks for the feedback. Interesting idea. So to group related tests, you'd use submodules instead of test suites?

I think unit tests based on a unit test framework are useful in that they provide a convention for when a test is considered failed or not, and how to run the tests and report their results. Of course, this requires that the convention for the test framework is rather strong (compare https://xkcd.com/927/ otherwise :-) ).

**Comment by mfelleisen** [comment]
Yes. The “downside" is that you need to specify the name of the test submodules.  (It’s like saying writing down types in s statically but implicitly typed PL, say ML, is a downside.) 

This trick also allows me to build up examples separately from tests so that I load tests only once when I want to run all of them. I have been meaning to write a short essay on that.

**Comment by sschwarzer** [comment]
> The “downside" is that you need to specify the name of the test submodules.

I have no problem with this. :-) I guess in most cases this is actually easier than specifying a string (pattern) for the test suite/case descriptions.

Is it possible to specify nested module names with `raco test -s ...`? In that case I could have a `test` submodule (to run all tests) and in the `test` module submodules corresponding to my former test suites.

**Comment by mfelleisen** [comment]
No. Use this organization instead: 

```
#lang racket

(require rackunit)

(define (f x) x) 

(module+ f-tests
  (check-equal? (f 1) 1))

(module+ test
  (require (submod ".." f-tests)))
```

Then run 

```
$ raco test -s f-tests foo.rkt
raco test: (submod "foo.rkt" f-tests)
raco test: @(test-responsible '(matthias))
1 test passed
$  raco test -s test foo.rkt
raco test: (submod "foo.rkt" test)
raco test: @(test-responsible '(matthias))
1 test passed
```

instead. This let’s you choose between “all tests” and “some tests” as needed;

**Comment by alex-hhh** [comment]
I wrote a package that does mostly what you want, except it does not look at the command line, instead it changes the call to `run-tests`, for example, you can either exclude individual tests or test suites or include only a specific test or test suite:

```racket
(run-tests
  #:package "my-package"
  #:results-file "my-package-test-results.xml"
  ;  Run everything except the second test case
  #:exclude '(("A Test Suite" "Second Test Case"))
  a-test-suite)

(run-tests
  #:package "my-package"
  #:results-file "my-package-test-results.xml"
  #:only '(("A Test Suite" "First Test Case"))
  a-test-suite)
```

The rackunit package is flexible enough to allow writing different ways of running tests, this is great, since different people have different requirements.  In particular, my package prints out the name of all the tests it runs or skips, which others might find too verbose.

You can use the package directly, or you can look at the source code to write a test runner that suits your needs:

https://pkgs.racket-lang.org/package/al2-test-runner

https://github.com/alex-hhh/al2-test-runner
